### PR TITLE
Feature/authenticated url

### PR DESCRIPTION
**Pull request by theprogrammist**
raw milestone for checkup

usage:
// create object instance for Api.
$api = new Api(UC_PUBLIC_KEY, UC_SECRET_KEY, null, UC_CNAME, null, null, 3600, UC_CHAME_SECRET);

$f1 = $api->getFile('https://ucarecdn.com/707eda96-77e7-4e98-9bea-dd1215a35078/-/preview/1440x1000/-/setfill/ffffff/-/format/webp/-/progressive/no/-/gamma/10/');
$f2 = $api->getFile('707eda96-77e7-4e98-9bea-dd1215a35078/-/preview/1440x1000/-/setfill/ffffff/-/format/webp/-/progressive/no/-/gamma/10/');
$f3 = $api->getFile('707eda96-77e7-4e98-9bea-dd1215a35078');

var_dump($f1->getUrl());
var_dump($f2->getUrl());
var_dump($f3->getUrl());

### CODE REVIEW

File: src/Uploadcare/Api.php [hunk 1]
```diff
@@ -139,6 +147,9 @@ class Api
      * @param string $cdn_protocol CDN Protocol
      * @param integer $retry_throttled Retry throttled requests this number of times
      * @param int $lifetime Secure signature expire time in seconds for signed uploads.
+     * @param string $cname_secret secret key for authentication token generation
+     * @param string $cdn_enc_algo algorithm, one of 'sha256','sha1','md5'
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Api.php]: we don't need this flexibility, just hardcoding sha256 is good enough.

File: src/Uploadcare/Api.php [hunk 2]
```diff
@@ -148,7 +159,10 @@ public function __construct(
         $cdn_host = null,
         $cdn_protocol = null,
         $retry_throttled = null,
-        $lifetime = 0
+        $lifetime = 0,
+        $cname_secret = '',
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Api.php]:
```suggestion
        $cdn_secret_token = '',
```

File: src/Uploadcare/Api.php [hunk 3]
```diff
@@ -33,6 +34,13 @@ class Api
      */
     private $api_host = 'api.uploadcare.com';
 
+    /**
+     * AuthenticatedUrl concrete realization
+     *
+     * @var Authenticate\AuthenticatedUrlInterface
+     */
+    public $authenticatedUrl;
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Api.php]: why do we want this public?

File: src/Uploadcare/Api.php (hunk 2 above)
Comment (theprogrammist) [reply src/Uploadcare/Api.php]: Это ж не токен, токен на выходе в url это key))

File: src/Uploadcare/Api.php (hunk 3 above)
Comment (theprogrammist) [reply src/Uploadcare/Api.php]: strategy pattern, we use it in File class to make file->getUrl polymorphic transparent

File: tests/Uploadcare/Authenticate/AkamaiAuthenticatedUrlTest.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        $authenticatedUrl = new AkamaiAuthenticatedUrl($secretKey, $expireTimeInSeconds);
+
+        $dateTime = new DateTime();
+        $dateTimeTimestamp = $dateTime->getTimestamp();
+        $expireTimestamp = $dateTimeTimestamp + $expireTimeInSeconds;
+
+        $this->assertEquals('exp=' . $expireTimestamp . $authenticatedUrl->getFieldDelimiter(), $authenticatedUrl->getExprField());
+    }
+}
```
Comment (dmitry-mukhin) [inline tests/Uploadcare/Authenticate/AkamaiAuthenticatedUrlTest.php]: please configure your editor to add a new line at EOF

File: tests/ApiTest.php [hunk 5]
```diff
@@ -24,6 +27,7 @@ class ApiTest extends TestCase
     public function setUp()
     {
         $this->api = new Api(UC_PUBLIC_KEY, UC_SECRET_KEY);
+        $this->authenticatedApi = new Api(UC_PUBLIC_KEY, UC_SECRET_KEY, null, UC_CNAME, null, null, 3600, UC_CHAME_SECRET);
```
Comment (dmitry-mukhin) [inline tests/ApiTest.php]:
```suggestion
        $this->authenticatedApi = new Api(UC_PUBLIC_KEY, UC_SECRET_KEY, null, UC_CNAME, null, null, 3600, UC_CNAME_SECRET);
```

File: src/Uploadcare/File.php [hunk 6]
```diff
@@ -77,6 +85,12 @@ public function __construct($uuid_or_url, Api $api, $data = false)
                 $this->default_effects = $this->cached_data['default_effects'];
             }
         }
+
+        if ($api->cdn_secret_token) {
+            if ($api->cdn_provider === 'akamai') {
+                $this->authenticatedUrl = new AkamaiAuthenticatedUrl($this->api->cdn_secret_token, $this->api->lifetime);
+            }
```
Comment (dmitry-mukhin) [inline src/Uploadcare/File.php]: else, throw something, e.g. Exception("Not Implemented")

File: src/Uploadcare/File.php [hunk 7]
```diff
@@ -33,6 +34,13 @@ class File
      */
     private $api = null;
 
+    /**
+     * AuthenticatedUrl concrete realization
+     *
+     * @var Authenticate\AuthenticatedUrlInterface
+     */
+    public $authenticatedUrl;
```
Comment (dmitry-mukhin) [inline src/Uploadcare/File.php]:
```suggestion
    private $authenticatedUrl;
```

File: tests/ApiTest.php [hunk 8]
```diff
@@ -464,6 +468,29 @@ public function testUploadFromStringWithCustomFilename()
...
 
+    /**
+     * Test uploading from path
+     */
+    public function testAuthenticatedUrls()
+    {
+        try {
+            $file = $this->authenticatedApi->uploader->fromPath(dirname(__FILE__).'/test.jpg', 'image/jpeg', 'rename.jpg');
+        } catch (Exception $e) {
+            $this->fail('We get an unexpected exception trying to upload from path: '.$e->getMessage());
```
Comment (dmitry-mukhin) [inline tests/ApiTest.php]: if we're here, it's not a test fail, it's an error.

File: tests/ApiTest.php [hunk 9]
```diff
@@ -464,6 +468,29 @@ public function testUploadFromStringWithCustomFilename()
...
+        }
+
+        $dargs = array("ssl" => array("verify_peer" => false, "verify_peer_name" => false), "http" => array('timeout' => 60, 'user_agent' => 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.9) Gecko/20071025 Firefox/3.0.0.1'));
+
+        try {
+            $this->assertEquals(
+                file_get_contents($file->getUrl(), false, stream_context_create($dargs)),
+                file_get_contents(dirname(__FILE__).'/test.jpg')
+            );
```
Comment (dmitry-mukhin) [inline tests/ApiTest.php]: simple HTTP status check would be sufficient here.

File: tests/ApiTest.php [hunk 10]
```diff
@@ -464,6 +468,29 @@ public function testUploadFromStringWithCustomFilename()
...
+        } catch (Exception $e) {
+            $this->fail('Error while compare file downloaded by authenticated url and the original : '.$e->getMessage());
```
Comment (dmitry-mukhin) [inline tests/ApiTest.php]: you will catch a lot of other things here, and the error message will be confusing and mask real reasons (e.g. for network errors).

File: tests/config.php [hunk 11]
```diff
@@ -2,3 +2,5 @@
 define('UC_PUBLIC_KEY', 'demopublickey');
 define('UC_SECRET_KEY', 'demoprivatekey');
 date_default_timezone_set('UTC');
+define('UC_CNAME', 'railsmuffin.ucarecdn.com');
```
Comment (dmitry-mukhin) [inline tests/config.php]: remove this

File: tests/ApiTest.php (hunk 5 above)
Comment (theprogrammist) [reply tests/ApiTest.php]: Why here is no facepalm emoticon?))

**Comment by dmitry-mukhin** [comment]
"Couldn't resolve host 'https'"

File: src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        $bin = "";
+        $i = 0;
+        do {
+            $bin .= chr(hexdec($str{$i} . $str{($i + 1)}));
+            $i += 2;
+        } while ($i < strlen($str));
+        return $bin;
+    }
+
+    public function getExprField()
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php]: should this be public?

File: src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return 'exp=' . $this->expire . $this->fieldDelimiter;
+    }
+
+    /**
+     * @param $url string
+     *
+     * @return string
+     */
+    public function getAclField($url)
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php]: should this be public?

File: src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return 'acl=' . $url . $this->fieldDelimiter;
+    }
+
+    public function getFieldDelimiter()
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php]: should this be public?

File: src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return $this->fieldDelimiter;
+    }
+
+    public function getKey()
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Authenticate/AkamaiAuthenticatedUrl.php]: should this be public?

File: src/Uploadcare/File.php [hunk 16]
```diff
@@ -87,6 +87,10 @@ public function __construct($uuid_or_url, Api $api, $data = false)
         }
 
         if ($api->cdn_secret_token) {
+            if(empty($api->cdn_host)) {
+                throw new \Exception('CDN Host must not be emty');
```
Comment (dmitry-mukhin) [inline src/Uploadcare/File.php:91]:
```suggestion
                throw new \Exception('CDN Host must not be empty');
```

**Pull request merged**